ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java

```diff
@@ -323,16 +323,16 @@ static void sendReply(List<CompletableFuture<DataStreamReply>> remoteWrites,
     ctx.writeAndFlush(builder.build());
   }
 
-  private CompletableFuture<Void> startTransaction(StreamInfo info, DataStreamRequestByteBuf request, long bytesWritten,
+  private void startTransaction(StreamInfo info, DataStreamRequestByteBuf request, long bytesWritten,
```

Comment:
Let's keep it async and sync it outside.
```
+++ b/ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java
@@ -409,7 +409,7 @@ public class DataStreamManagement {
           } else if (close) {
             if (info.isPrimary()) {
               // after all server close stream, primary server start transaction
-              startTransaction(info, request, bytesWritten, ctx);
+              startTransaction(info, request, bytesWritten, ctx).join();
             } else {
               sendReply(remoteWrites, request, bytesWritten, info.getCommitInfos(), ctx);
             }
```

Reply:
Has been modified